README
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.32
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "SEH 1"
.TH SEH 1 "2009-05-31" "perl v5.8.8" "User Contributed Perl Documentation"
.SH "NAME"
Exception::SEH \- rich try/catch/finally semantics without source filtering
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 10
\&        use Exception::SEH;
\&        try {
\&                <...> or die YourException;
\&        } catch (SomeException $e) {
\&                <catch code>
\&        } catch (YourException $e where { $_->param > 10 }) {
\&                log($e->message);
\&        } finally {
\&                <cleanup code>
\&        }
.Ve
.SH "FEATURES"
.IX Header "FEATURES"
.IP "\(bu" 2
Rich catch syntax
.IP "\(bu" 2
No need for trailing ; after try/catch/finally sequence
.IP "\(bu" 2
return from try/catch \- returns from subroutine, not just try/catch block
.IP "\(bu" 2
finally called in all cases (even after return)
.IP "\(bu" 2
@_ available for try block
.IP "\(bu" 2
Nested try/catch
.IP "\(bu" 2
Not a source filter (magic for syntax is provided by Devel::Declare)
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
.IP "try \s-1BLOCK\s0" 4
.IX Item "try BLOCK"
Evaluates \s-1BLOCK\s0, allowing any exceptions inside it to be caught and processed.
.IP "catch \s-1BLOCK\s0" 4
.IX Item "catch BLOCK"
.PD 0
.ie n .IP "catch ([[\s-1CLASS\s0] $VAR] [where \s-1BLOCK\s0]) \s-1BLOCK\s0" 4
.el .IP "catch ([[\s-1CLASS\s0] \f(CW$VAR\fR] [where \s-1BLOCK\s0]) \s-1BLOCK\s0" 4
.IX Item "catch ([[CLASS] $VAR] [where BLOCK]) BLOCK"
.PD
Catches and processes (using given \s-1BLOCK\s0) exception that (optionally) matches
some conditions. Rules to determine catch handler are following:
.RS 4
.IP "*" 2
Search is performed from beginning to end of 'catches' list
.IP "*" 2
First matched block is executed and the search is stopped
.IP "*" 2
If exception is not an object, then any condition with \s-1CLASS\s0 definition is skipped
.IP "*" 2
If condition has \s-1CLASS\s0, then \fB$@\-\fRisa(\s-1CLASS\s0)> is checked for true value
.IP "*" 2
If condition has 'where \s-1BLOCK\s0' part, then \s-1BLOCK\s0 is checked for true result, with \fB$_\fR assigned to caught exception
.IP "*" 2
catch \s-1BLOCK\s0, catch () \s-1BLOCK\s0, catch ($e) \s-1BLOCK\s0 are all the same and captures all possible exceptions
.IP "*" 2
If no condition matches throwed exception, then it's re-thrown (after \f(CW\*(C`finally\*(C'\fR is done)
.RE
.RS 4
.Sp
Inside \s-1BLOCK\s0 after \f(CW\*(C`catch\*(C'\fR exception is available as \fB$@\fR (always) and also as \f(CW$VAR\fR (lexical to \s-1BLOCK\s0),
if it was specified in \f(CW\*(C`catch\*(C'\fR signature.
.RE
.IP "finally \s-1BLOCK\s0" 4
.IX Item "finally BLOCK"
Finally block is always executed after either \f(CW\*(C`try\*(C'\fR or \f(CW\*(C`catch\*(C'\fR. Exception in it is available as \fB$@\fR.
.Sp
Note: if exception was generated inside a \f(CW\*(C`catch\*(C'\fR handler, than in \fB$@\fR would be new one, not
original from a \f(CW\*(C`try\*(C'\fR block.
.Sp
Note: \f(CW\*(C`exit\*(C'\fR and unhandled signals leads to immidiate program termination (latter even don't allow
global destruction to happen) \- in these cases, \f(CW\*(C`finally\*(C'\fR is \fBnot\fR called.
.IP "return \s-1EXPRESSION\s0" 4
.IX Item "return EXPRESSION"
\&\f(CW\*(C`return\*(C'\fR behaviour is modifiyed, so when you call it inside \f(CW\*(C`try\*(C'\fR/\f(CW\*(C`catch\*(C'\fR, it returns not from 
\&\f(CW\*(C`try\*(C'\fR block, but from the sub \f(CW\*(C`try\*(C'\fR was inside.
.Sp
Note: this does not apply to finally block.
.Sp
Note: if both \f(CW\*(C`catch\*(C'\fR and \f(CW\*(C`try\*(C'\fR uses \f(CW\*(C`return\*(C'\fR, than value from last \f(CW\*(C`return\*(C'\fR is used (but you
may ommit \f(CW\*(C`return\*(C'\fR in \f(CW\*(C`catch\*(C'\fR block, then return value is left unchanged).
.Sp
Note: argument for return is executed in the context of sub above \f(CW\*(C`try\*(C'\fR.
.Sp
Example:
.Sp
.Vb 7
\&        sub test{
\&                try{
\&                        return map { $_ * 2 } @_;
\&                }
\&        }
\&        $a = test(1, 2, 3);     #$a = 3
\&        @a = test(1, 2, 3);     #@a = (2, 4, 6)
.Ve
.IP "@_" 4
.IX Item "@_"
Inisde \f(CW\*(C`try\*(C'\fR, you have full read/write access to surronding sub's \f(CW@_\fR. 
.Sp
Note: inside \f(CW\*(C`catch\*(C'\fR/\f(CW\*(C`finally\*(C'\fR, you have only read-only access to \f(CW@_\fR.
.SH "CONFIGURATION"
.IX Header "CONFIGURATION"
When importing Exception::SEH to your module, you can specify
one or more of the following keywords: \fB\-nosig\fR \fB\-noret\fR \fB\-safetry\fR, which change default
behaviour in some way. These options are tracked on per-package basis.
.PP
.Vb 1
\&        use Exception::SEH -nosig, -safetry;
.Ve
.IP "\-nosig" 4
.IX Item "-nosig"
By default, Exception::SEH doesn't play with \f(CW$SIG\fR{_\|_DIE_\|_}, so your handler
is normally called. But if you need to suppress it inside try/catch/finally blocks,
you can import \f(CW\*(C`\-nosig\*(C'\fR.
.Sp
Note: if exception is re-thrown after \f(CW\*(C`finally\*(C'\fR block, _\|_DIE_\|_ handler is
always suppressed for that moment (so it is called only once
for each exception \- at first occurence).
.IP "\-safetry" 4
.IX Item "-safetry"
This options changes behaviour for uncaught exceptions after \f(CW\*(C`try\*(C'\fR block.
By default, they're re\-thrown, but with this option enabled \f(CW\*(C`try\*(C'\fR blocks are always safe \- just like \f(CW\*(C`eval\*(C'\fR.
.Sp
Note: this doesn't refer to exceptions occured in a \f(CW\*(C`catch\*(C'\fR and \f(CW\*(C`finally\*(C'\fR blocks.
.IP "\-noret" 4
.IX Item "-noret"
Tells Exception::SEH not to install \f(CW\*(C`return\*(C'\fR hook. This makes \f(CW\*(C`try\*(C'\fR blocks behave like
evals, and context for \f(CW\*(C`return\*(C'\fR is determined by context of \f(CW\*(C`try\*(C'\fR, not by context of sub,
containing \f(CW\*(C`try\*(C'\fR.
.SH "DIAGNOSTICS"
.IX Header "DIAGNOSTICS"
All compile-time discovered errors are raised through \f(CW\*(C`Carp::croak\*(C'\fR
with (hopefully) meaningful description.
.SH "CAVEATS"
.IX Header "CAVEATS"
Currently, three subroutines \- try, catch, finally \- are injected into caller's namespace.
.PP
Currently, it's impossible to use any other module that hooks \fB\s-1OP_RETURN\s0\fR \s-1OP\s0 on \s-1XS\s0 level \-
results are unpredictable. This may be fixed in future releases.
.PP
Currently, it's probably unsafe to place \f(CW\*(C`try\*(C'\fR/\f(CW\*(C`catch\*(C'\fR blocks around \f(CW\*(C`use\*(C'\fR/\f(CW\*(C`catch\*(C'\fR.
This needs further testing, and may be fixed in future releases.
.PP
Currently, due to way caller is detected, if you place try/catch/finally inside \f(CW\*(C`eval\*(C'\fR block,
\&\f(CW\*(C`return\*(C'\fR wouldn't work as expected. This will be fixed in future releases.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
TryCatch
.PP
Similar semantics, but it takes the whole Moose with itself \- often too much weight.
Currently lacks finally block and \fB@_\fR access inside try{}. But still, decent choice.
.PP
Exception::Class::TryCatch
.PP
Ugly syntax. Only Exception::Class\-derived exceptions.
.PP
Exception::Class::TCF
.PP
Better syntax. Lacks exception lexical in try block, returns from try{} block.
Still only Exception::Class\-derived exceptions.
.PP
Error::TryCatch
.PP
Source filter. No finally{} call after \fIreturn()\fR from try.
.PP
Error
.PP
\&\*(L"Error's syntactic sugar tends to break.\*(R" (c)
.SH "AUTHOR"
.IX Header "AUTHOR"
Sergey Aleynikov <sergey.aleynikov@gmail.com>
.SH "LICENSE"
.IX Header "LICENSE"
Copyright (c) 2009 by Sergey Aleynikov.
This program is free software; you can redistribute it and/or modify it under the same terms as Perl itself.
.PP
Some part for parser are taken from Devel::Declare::Context::Simple, (c) Rhesa Rozendaal (?).
.PP
Idea how to set up return hook \- TryCatch, (c) Ash Berlin.


Changes
Revision history for Exception::SEH.

0.01	2009-05-28
		- initial release

0.0101	2009-05-28
		- removed code left from debugging

0.0102	2009-05-31
		- updated documentation
		- propagade @_ to catch and finally

0.0103	2009-06-08
		- sometimes wrong options could be picked
		  when multiply packages were defined in the same file

0.0104	2009-06-08
		- forgot to update Changes

0.0105	2009-06-20
		- fixed compilation issues for threaded perls
		- added META.yml

0.0106	2009-08-30
		- fix possible coredump for 5.8.8 with usemymalloc=no
		- private release

0.02	2009-11-14
		- rewritten parser for catch/finally
		- parser does not call exit() any more
		- parser error handling differs from 0.01,
			now all parse errors are printed to STDERR and follows some perl message
			this also affect string evals - sometime they get only perl message
			and sometime a correct one
		- speedup of many-catches code
		- more effort is taken to catch errors inside 'catch' correctly

0.0201	2009-11-21
		- explicitly forbid compilation under Win32,
			as 'return' hooking fails there

0.0202	2011-04-26
		- do not use core non-exported symbol


Makefile.PL
use strict;
use inc::Module::Install;

die "OS unsupported\n" if $^O =~ /MSWin/;

perl_version '5.008001';
name 'Exception-SEH';

abstract_from 'lib/Exception/SEH.pod';
author 'Sergey Aleynikov <sergey.aleynikov@gmail.com>';
version_from 'lib/Exception/SEH.pm';
license 'perl';

test_requires 'Test::More';
test_requires 'Test::Exception';
test_requires 'Test::Warn';

configure_requires 'B::Hooks::OP::PPAddr' => '0.02';
configure_requires 'B::Hooks::OP::Check' => '0.15';
configure_requires 'ExtUtils::Depends';

requires 'Carp';
requires 'XSLoader';
requires 'Scalar::Util';
requires 'Scope::Upper' => '0.07';
requires 'Devel::Declare' => '0.005';
requires 'B::Hooks::EndOfScope' => '0.08';

eval 'require ExtUtils::Depends' or die 'You must install ExtUtils::Depends, B::Hooks::OP::PPAddr, B::Hooks::OP::Check before running install';
my $pkg = ExtUtils::Depends->new('TryCatch', 'B::Hooks::OP::PPAddr', 'B::Hooks::OP::Check');
makemaker_args(
    $pkg->get_makefile_vars,
);

system 'pod2text lib/Exception/SEH.pod > README';

WriteAll;


lib/Exception/SEH/Parser.pm
#!/usr/bin/perl
package Exception::SEH::Parser;

use strict;

use Carp ();
use Devel::Declare ();
use B::Hooks::OP::PPAddr;
use Scope::Upper qw(EVAL);

sub DEBUG() { 0 }

sub INITIAL()	{ -1 }
sub TRY()		{ 0 }
sub CATCH()		{ 1 }
sub FINALLY()	{ 2 }

our $VERSION = '0.01003';

sub new{
	my ($class, $offset) = @_;

	print STDERR "new called at $offset\r\n" if DEBUG;
	bless {
		offset	=> $offset,
	}, $class;
}

#err handler

sub panic{
	my ($self, $err) = @_;

	if (EVAL > 0){
		Carp::croak $err;
	}else{
		print STDERR 'Exception::SEH - ', $err, "\n\n";
		die;
	}
}

#token manip

sub skip_word{
	my $self = shift;

	print STDERR "skip_word called at $self->{offset}\r\n" if DEBUG;

	if (my $len = Devel::Declare::toke_scan_word($self->{'offset'}, 1)) {
		$self->{'offset'} += $len;
	}
}

sub get_word{
	my $self = shift;

	print STDERR "get_word called at $self->{offset}\r\n" if DEBUG;

	if (my $len = Devel::Declare::toke_scan_word($self->{'offset'}, 1)) {
		return substr(Devel::Declare::get_linestr(), $self->{'offset'}, $len);
	}
	return '';
}

sub skip_spaces{
	my $self = shift;

	print STDERR "skip_spaces called at $self->{offset}\r\n" if DEBUG;

	$self->{'offset'} += Devel::Declare::toke_skipspace($self->{'offset'});
}

sub get_symbols{
	my ($self, $len) = @_;

	print STDERR "get_symbols called at $self->{offset} for $len\r\n" if DEBUG;

	return substr(Devel::Declare::get_linestr(), $self->{'offset'}, $len);
}


sub extract_args{
	my $self = shift;

	print STDERR "extract_args called at $self->{offset}\r\n" if DEBUG;

	my $linestr = Devel::Declare::get_linestr();
	if (substr($linestr, $self->{'offset'}, 1) eq '(') {
		my $length = Devel::Declare::toke_scan_str($self->{'offset'});
		my $proto = Devel::Declare::get_lex_stuff();
		Devel::Declare::clear_lex_stuff();

		$linestr = Devel::Declare::get_linestr();
		if (
			$length < 0
				||
			$self->{'offset'} + $length > length($linestr)
				||
			$self->{'offset'} < 0
		){
			$self->panic("Unbalanced text supplied as catch argument");
		}
		substr($linestr, $self->{'offset'}, $length) = '';
		Devel::Declare::set_linestr($linestr);

		return $proto;
	}
	return '';
}

#injectors

sub inject{
	my ($self, $string) = @_;

	$self->substitute($string, 0);
}

sub cutoff{
	my ($self, $len) = @_;

	$self->substitute('', $len);
}

sub substitute{
	my ($self, $string, $replace_len) = @_;

	print STDERR "inject called at $self->{offset} for '$string'\r\n" if DEBUG;

	my $linestr = Devel::Declare::get_linestr;
	if (
		$self->{'offset'} > length($linestr)
				||
		$self->{'offset'} < 0
	){
		$self->panic("Parser tried to inject data outside program source, stopping");
	}
	substr($linestr, $self->{'offset'}, $replace_len) = $string;
	Devel::Declare::set_linestr($linestr);

	$self->{'offset'} += length($string);
}

sub inject_if_block{
	my ($self, $inject) = @_;

	print STDERR "inject_if_block called at $self->{offset} for '$inject'\r\n" if DEBUG;

	if ($self->get_symbols(1) eq '{'){
		$self->{'offset'} += 1;
		$self->inject($inject);
	}else{
		$self->panic('Code block expected');
	}
}

sub get_injector{
	my ($self, $func, @args) = @_;

	return " BEGIN { $func(".join(',', map { "'$_'" } @args).") } ";
}

1;

=head1 NAME

Exception::SEH::Parser - parses source for L<Exception::SEH> and is not intended for external use.

=head1 AUTHOR

Copyright (c) 2009 by Sergey Aleynikov.
This program is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

=cut


lib/Exception/SEH.pm
#!/usr/bin/perl
package Exception::SEH;

use strict;
use 5.008001;

use Carp ();
use Devel::Declare ();
use B::Hooks::EndOfScope;
use Exception::SEH::Parser;
use Scalar::Util qw(blessed);
use Scope::Upper qw(unwind want_at reap :words);

use XSLoader;

BEGIN{
	no strict 'refs';
	foreach (qw(INITIAL TRY CATCH FINALLY DEBUG)) {
		*{'Exception::SEH::'.$_} = *{'Exception::SEH::Parser::'.$_};
	}
}

our $VERSION = '0.0202';
$Carp::Internal{'Exception::SEH'}++;
$Carp::Internal{'Devel::Declare'}++;

our %OPTS = ();
our $params = [];
our $need_unwind = 0;
our $hook_nested_level = 0;
our $parse_catch_called = 0;
our $return_hook_id = undef;

XSLoader::load(__PACKAGE__, $VERSION);

sub import{
	my $class = shift;
	my $caller = caller;

	my $cur_opts = $OPTS{$caller} = {
		-nosig	 => 0,
		-noret	 => 0,
		-safetry => 0,
	};

	foreach (@_) {
		if (exists $cur_opts->{$_}){
			$cur_opts->{$_} = 1;
		}else{
			Carp::croak "Tried to set non-existent option: $_";
		}
	}

	#try
	Devel::Declare->setup_for(
		$caller,
		{ try => { const => \&parse_try } }
	);

	no strict 'refs';

	#this'd be shadowed
	*{$caller.'::try'} = \&try;

	#these are never directly called
	*{$caller.'::finally'}	= sub (;&) { Carp::croak 'Found finally without a try{} block' };
	*{$caller.'::catch'}	= sub (;&) { Carp::croak 'Found catch without a try{} block' };
}

sub parse_try{
	my $parser = Exception::SEH::Parser->new($_[1]);

	if ((my $token = $parser->get_word()) ne 'try'){
		return;
	}

	$parser->skip_word();
	$parser->skip_spaces();
	if ($parser->get_symbols(2) eq '=>'){
		return;
	}

	$parser->inject('(\@_, sub');

	$return_hook_id = Exception::SEH::XS::install_return_op_check()
		if !$OPTS{Devel::Declare::get_curstash_name}->{'-noret'} && !$return_hook_id;
	$hook_nested_level++;

	$parser->inject_if_block($parser->get_injector('Exception::SEH::aftercheck', TRY));
	$parser->inject('*_ = $Exception::SEH::params;');
}

sub aftercheck{
	my $prev_item_type = shift;

	on_scope_end {
		if (
			$return_hook_id
				&&
			--$hook_nested_level <= 0
		){
			Exception::SEH::XS::uninstall_return_op_check($return_hook_id);
			$return_hook_id = undef;
		}

		my $parser = Exception::SEH::Parser->new(Devel::Declare::get_linestr_offset);

		$parser->skip_spaces();
		my $token = $parser->get_word();

		if ($token eq 'catch'){
			parse_catch($parser, $prev_item_type);
		}elsif ($token eq 'finally'){
			parse_finally($parser, $prev_item_type);
		}else{
			$parser->inject(');');
		}
	}
}

sub parse_finally{
	my ($parser, $prev_item_type) = @_;

	if ($prev_item_type == FINALLY){
		$parser->panic('Found finally block more than once after single try{}');
	}

	$parser->cutoff(length('finally'));
	$parser->inject(', sub ');
	$parser->skip_spaces();
	$parser->inject_if_block($parser->get_injector('Exception::SEH::aftercheck', FINALLY));
}

sub parse_catch{
	my ($parser, $prev_item_type) = @_;

	if ($prev_item_type == FINALLY){
		$parser->panic('Found catch block again after finally');
	}

	$parser->cutoff(length('catch'));
	$parser->inject(', ');
	$parser->skip_spaces();

	my $err_var = undef;

	if ($parser->get_symbols(1) eq '('){
		my $args = $parser->extract_args();

		if (length($args)){
			my ($type) = $args =~ /\G\s*(?!where\W)([a-zA-Z0-9_:]+)\s*/gcs;
			($err_var) = $args =~ /\G\s*\$(\w+)\s*/gcs;

			if ($type && !$err_var){
				$parser->panic("Found exception class definition, but no context var while parsing catch definition");
			}

			if ($type){
				$parser->inject(" '$type', ");
			}else{
				$parser->inject(" undef, ");
			}

			my ($where_present, $where) = $args =~ /\G\s*(where)\s*(.*?)\s*$/gcs;
			if ($where_present && $where !~ /^{.*}$/){
				$parser->panic('"Where" must be followed by a block inside catch definition');
			}

			if ($where){
				$parser->inject(" sub $where");
			}else{
				$parser->inject(" undef");
			}

			if (pos($args) != length($args)){
				$parser->panic('Found junk inside catch definition');
			}

		}else{
			$parser->inject(' undef, undef');
		}

	}else{
		$parser->inject(' undef, undef');
	}

	$parser->inject(', sub ');
	$return_hook_id = Exception::SEH::XS::install_return_op_check()
		if !$OPTS{Devel::Declare::get_curstash_name}->{'-noret'} && !$return_hook_id;
	$hook_nested_level++;

	$parser->skip_spaces();
	$parser->inject_if_block($parser->get_injector('Exception::SEH::aftercheck', CATCH));
	$parser->inject('my $'.$err_var.' = $@;') if $err_var;
}

#==========##==========##==========#

sub try($&@) {
	my $opts = $OPTS{scalar caller};
	local $SIG{__DIE__} = 'DEFAULT' if $opts->{'-nosig'};

	#for unwind
	my $is_top_try = 0;

	if (!defined(EVAL) || EVAL != CALLER 2){	#we're inside top try
		$is_top_try = 1;
	}
	local $params = shift;
	my $code = shift;

	my $finally = (scalar @_ % 3 == 0 ? undef : pop @_);
	my $catch_fail = 0;
	my $exception_caught = 1;

	my $cx = $is_top_try ? UP SUB : EVAL;
	my $context = $opts->{'-noret'} ? wantarray : want_at($cx);

	$@ = undef;
	my @result;
	if ($context){
		@result = eval { $code->() }
	}elsif(defined($context)){
		$result[0] = eval { $code->() }
	}else{
		eval { $code->() }
	}

	my $err = $@;
	$exception_caught = 0 if $err;

	if ($err && scalar @_){
		my $checked = 0;

		eval{
			my $err_blessed = blessed($err);

			my $pos = -3;
			while ($pos + 3 < scalar @_) {
				$pos += 3;
				my ($type, $where, $handler) = @_[$pos..$pos+2];

				if (
					!$type
						||
					 $err_blessed
						&&
					$err->isa($type)
				){
					if (defined $where){
						local $_ = $err;
						next if !$where->();
					}

					$exception_caught = 1;
					my $_need_unwind = $need_unwind;
					$need_unwind = 0;
					my @_result;

					eval {
						$@ = $err;
						if ($context){
							@_result = $handler->(@$params);
						} elsif (defined($context)) {
							$_result[0] = $handler->(@$params);
						} else {
							$handler->(@$params);
						};
						1;
					} or do {
						$err = $@;
						$catch_fail = 1;
					};

					if ($need_unwind){
						@result = @_result;
					}else{
						$need_unwind = $_need_unwind;
					}
					last;
				}
			}
			1;
		} or do {
			$err = $@;
			$catch_fail = 1;
		}
	}

	if (defined $finally){
		$@ = $err;
		$finally->(@$params);
	}

	if ($catch_fail || (!$exception_caught && !$opts->{'-safetry'})){
		local $SIG{__DIE__} = 'DEFAULT';
		die $err;
	}

	if($opts->{'-noret'} || !$need_unwind){
		#print STDERR "normal return\n";
		$need_unwind = 0;
		return wantarray ? @result : $result[0];

	}elsif($need_unwind){
		$need_unwind = 0 if $is_top_try;
		unwind +($context ? @result : $result[0]) => $cx;

	}else{
		Carp::croak 'Cannot determine return point';
	}
}

1;

=head1

Copyright (c) 2009 by Sergey Aleynikov.
This program is free software; you can redistribute it and/or modify it under the same terms as Perl itself.

=cut
